Fixups to the unit enablement logic

keszybz opened the pull request.

yuwata commented:
I think it is better to add more test cases for the fixed issues.

poettering commented on src/basic/unit-file.c:
```diff
@@ -261,6 +261,86 @@ static int directory_name_is_valid(const char *name) {
+                return log_warning_errno(r, "Failed to read symlink %s%s%s: %m",
+                                         dir, dir ? "/" : "", filename);
+
+        const bool is_abs = path_is_absolute(target);
```
"const bool"? please let's not start with this. Let's use it on pointers, but on local variables it's pointless clutter really.

poettering commented on src/basic/unit-file.c:
```diff
@@ -261,6 +261,86 @@ static int directory_name_is_valid(const char *name) {
+         * If yes, it's a linked unit file or mask, and we don't care about the target name.
+         * Let's just store the link source directly.
+         * If not, let's verify that it's a good symlink. */
+        char *tail = path_startswith_strv(simplified, search_path);
```
here otoh it would make sense to add a const, since the memory the ptr points to is foreign, and we should not change it

poettering commented on src/basic/unit-file.c:
```diff
@@ -261,6 +261,86 @@ static int directory_name_is_valid(const char *name) {
+        } else {
+                bool self_alias;
+
+                dst = basename(simplified);
```
if you touch this, maybe also use path_extract_filename()...

keszybz replied:
We know that the name is "well formed", and basename works nicely. It also avoid the need to pass a separate allocated pointer around and free it.

poettering replied:
I'd really prefer to add basename() to the list of functions that LGTM and coccinelle will complain/rewrite. The semantics are just too weird I think, and it's always good to have additional checks in place that the assumptions about "well-formedness" actually hold...

poettering commented on src/basic/unit-file.c:
```diff
@@ -270,19 +270,32 @@ static int unit_file_resolve_symlink(
+                        return r;
+
+                dir = _dir;
+                filename = basename(filename);
```
path_extract_filename()?

poettering commented on src/basic/unit-file.c:
```diff
@@ -261,6 +261,86 @@ static int directory_name_is_valid(const char *name) {
+                int dirfd,
+                const char *filename,
+                char **ret_simplified,
+                const char **ret_dst) {
```
the semantics of ret_dst are so mean, i.e. returning either a pointer to the pointer or to the simplified allocated buffer.

Afaics the caller doesn't bother about the `simplified` return parameter at all. Only about `dst`. Which it eventually duplicates to add to the set. It sounds much simpler if we'd just drop ret_simplified here, and return ret_dst as properly allocated string, that the caller can then just add to the set as-is, without eeding to duplicate it.

It's simpler and faster...

keszybz replied:
It's reworked along the suggested lines now.

bluca commented:
_holy patchset, Batman_

keszybz commented:
```
unit-config          FAIL non-zero exit status 1
storage              PASS
networkd-test.py     PASS
build-login          PASS
boot-and-services    PASS
udev                 PASS
unit-tests           FAIL badpkg
blame: https://salsa.debian.org/systemd-team/systemd.git#upstream-ci
badpkg: Test dependencies are unsatisfiable. A common reason is that your testbed is out of date with respect to the archive, and you need to use a current testbed or run apt-get update or use -U.
upstream             FAIL badpkg
```
So that looks at least partially unrelated, but `unit-config` is suspicious.

keszybz commented:
Yeah, so this would need to be adjusted. But let's see if we actually want to do this before changing the tests there.
```
======================================================================
FAIL: test_unit_alias_enable (__main__.EnableTests)
no sysv: enable unit with an alias
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest.OvTYXW/build.O10/systemd/debian/tests/unit-config", line 204, in test_unit_alias_enable
    self.assertEqual(os.readlink(l),
AssertionError: '../test_enable.service' != '/lib/systemd/system/test_enable.service'
- ../test_enable.service
+ /lib/systemd/system/test_enable.service


======================================================================
FAIL: test_unit_enable (__main__.EnableTests)
no sysv: enable unit
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest.OvTYXW/build.O10/systemd/debian/tests/unit-config", line 105, in test_unit_enable
    self.assertEqual(os.readlink(l),
AssertionError: '../test_enable.service' != '/lib/systemd/system/test_enable.service'
- ../test_enable.service
+ /lib/systemd/system/test_enable.service


======================================================================
FAIL: test_unit_enable_full_path (__main__.EnableTests)
systemctl enable a unit in a non-default path
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest.OvTYXW/build.O10/systemd/debian/tests/unit-config", line 356, in test_unit_enable_full_path
    self.assertEqual(os.readlink(enable_l), f.name)
AssertionError: '../tmpstnn9ebp.service' != '/tmp/tmpstnn9ebp.service'
- ../tmpstnn9ebp.service
? ^^
+ /tmp/tmpstnn9ebp.service
? ^^^^


======================================================================
FAIL: test_unit_sysv_alias_enable (__main__.EnableTests)
with sysv: enable unit with an alias
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest.OvTYXW/build.O10/systemd/debian/tests/unit-config", line 249, in test_unit_sysv_alias_enable
    self.assertEqual(os.readlink(l),
AssertionError: '../test_enable.service' != '/lib/systemd/system/test_enable.service'
- ../test_enable.service
+ /lib/systemd/system/test_enable.service


======================================================================
FAIL: test_unit_sysv_enable (__main__.EnableTests)
with sysv: enable unit
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/autopkgtest.OvTYXW/build.O10/systemd/debian/tests/unit-config", line 150, in test_unit_sysv_enable
    self.assertEqual(os.readlink(l),
AssertionError: '../test_enable.service' != '/lib/systemd/system/test_enable.service'
- ../test_enable.service
+ /lib/systemd/system/test_enable.service
```

poettering commented:
I cherry-picked a bunch of the initial commits into git main, since they should be riskless and make sense on their own. Please rebase

poettering commented on src/shared/install.c:
```diff
@@ -1344,68 +1344,21 @@ static int unit_file_load_or_readlink(
+         * When looking under root_dir, we can't expect /dev/ to be mounted,
+         * so let's see if the path is a (possibly dangling) symlink to /dev/null. */
+        if (null_or_empty_path(info->symlink_target) > 0 ||
+            (lp->root_dir && path_equal_ptr(path_startswith(info->symlink_target, lp->root_dir), "dev/null")))
```
this test expression looks like something that should be its own helper function, i.e. null_or_empty_path_with_root() or so.

poettering commented on src/basic/env-file.c:
```diff
@@ -414,30 +414,39 @@ static int load_env_file_push_pairs(
+                                r = free_and_replace(t[1], value);
+                        else
+                                r = free_and_strdup(t+1, "");
+                        goto fini;
```
what happened to `…sh`?

poettering replied:
in 99% of the cases we call the label we jump to for cleaning up "finish". But here you called it "fini", i.e. the "sh" somehow vanished. And I wondered where it went.

keszybz replied:
Oh, I see. No particular reason, I can make it "finish".

poettering commented on src/shared/install.c:
```diff
@@ -1152,7 +1158,8 @@ static int config_parse_also(
                 if (r < 0)
-                        return r;
+                        return log_syntax(unit, LOG_WARNING, filename, line, r,
+                                          "Failed to resolve unit name in Also \"%s\": %m", word);
```
suffix Also with =, i.e. Also=

poettering commented on src/shared/install.c:
```diff
@@ -1199,14 +1206,13 @@ static int config_parse_default_instance(
         if (r < 0)
-                return r;
+                return log_syntax(unit, LOG_WARNING, filename, line, r,
+                                  "Failed to resolve instance name in DefaultInstance \"%s\": %m", rvalue);
```
similar

poettering commented on src/shared/specifier.c:
```diff
@@ -67,6 +67,12 @@ int specifier_printf(const char *text, size_t max_length, const Specifier table[
+                                                 * which will return -ENOENT if the file is not present.
+                                                 * Let's translate this to -ENODATA to make the error more
+                                                 * understandable in the lower layers of the call stack. */
+                                                return -ENODATA;
```
hmm, shouldn't the individual specifier callbcks already convert that? Sounds weird to convert the errors so late... And shouldn't this be a bit more exotic errno? i.e. EUNATCH maybe?

poettering commented on src/shared/install.c:
```diff
@@ -49,6 +49,7 @@ typedef enum SearchFlags {
 typedef struct {
         OrderedHashmap *will_process;
         OrderedHashmap *have_processed;
+        UnitFileScope scope;
```
sounds more natural to put this first given it's initialized first?

poettering commented on src/shared/install.c:
```diff
@@ -2521,8 +2517,8 @@ int unit_file_add_dependency(
         _cleanup_(lookup_paths_free) LookupPaths lp = {};
-        _cleanup_(install_context_done) InstallContext c = {};
-        UnitFileInstallInfo *i, *target_info;
+        _cleanup_(install_context_done) InstallContext ctx = { .scope = scope };
```
why not add line breaks after the { and before the }?

keszybz replied:
I think that as long as only one field is initialized, it's OK to do it in one line. But I don't feel very strongly about this, I can change it if you think it looks better.

poettering commented on src/shared/install.c:
```diff
@@ -2633,7 +2630,7 @@ int unit_file_disable(
 
         _cleanup_(lookup_paths_free) LookupPaths lp = {};
-        _cleanup_(install_context_done) InstallContext c = {};
+        _cleanup_(install_context_done) InstallContext ctx = {};
```
initialization missing?

keszybz replied:
Ouch!

poettering commented on src/shared/install.c:
```diff
@@ -2734,8 +2731,8 @@ int unit_file_get_default(
         _cleanup_(lookup_paths_free) LookupPaths lp = {};
-        _cleanup_(install_context_done) InstallContext c = {};
-        UnitFileInstallInfo *i;
+        _cleanup_(install_context_done) InstallContext c = { .scope = scope };
```
line breaks?

poettering commented on src/shared/install.c:
```diff
@@ -2769,7 +2766,7 @@ int unit_file_lookup_state(
                 UnitFileState *ret) {
 
-        _cleanup_(install_context_done) InstallContext c = {};
+        _cleanup_(install_context_done) InstallContext c = { .scope = scope };
```
ditto (here and everywhere)

poettering commented on src/shared/specifier.c:
```diff
@@ -22,6 +22,7 @@
 #include "specifier.h"
 #include "string-util.h"
 #include "strv.h"
+#include "unit-file.h"
```
I'd really like to keep the specifier stuff indepdent of unit files. We use the specifier stuff all over the place after all.

hence, instead of this, why not define a flags field to pass in the `data` parameter, i.e.

```c
typedef enum SpecifierFlags {
        SPECIFIER_ROOT_USER = 1 << 0, /* force user/group resolution to return "root", "0" */
        SPECIFIER_FAIL_USER = 1 << 1, /* make user/group specifiers fail */
} SpecifierFlags;
```

and then add a little helper func that can translate the UnitFileScope to SpecifierFlags, so that you can easily set these flags based on the scope.

(Another idea would be to make these flags an extra parameter all specifier callbacks take, separate from `data`)

keszybz replied:
So… I don't like this approach. There'd be 1:1 mapping of the new enum and UnitFileScope. UnitFileScope is well established and widely used, and adding a new enum would just introduce a layer of abstraction that doesn't simplify anything. And I think it's OK to pass this as `data`. We don't use `data` for anything else, and for those specifiers, we're unlikely to ever do.

The part that is somewhat ugly are the conversions using `PTR_TO_INT` and `INT_TO_PTR`. But afaics, this part we can't get rid of.

poettering replied:
UnitFileScope is not well established outside of unit file handling. tmpfiles.d, sysusers.d and all the other tools that deal with specifiers cannot make sense of that...

poettering replied:
I mean, I would be more sympathetic to reusing UnitFileScope if it would be generalized, i.e. InvocationScope or AppScope or so, so that we can reasonably use it in all our tools, after all similar concepts kinda do make sense there.

poettering replied:
or tldr: I really don#t like the fact that specifier.h now includes unit-file.h.

poettering replied:
generalizing UnitFileScope as InvocationScope makes a ton of sense btw, we use similar concepts everywhere after all, i.e. when talking about D-Bus busses and stuff

keszybz replied:
New patch at the end now renames it to LookupScope.

poettering commented on src/shared/install.c:
```diff
@@ -2609,34 +2609,21 @@ int unit_file_add_dependency(
-                UnitFileScope scope,
-                UnitFileFlags file_flags,
-                const char *root_dir,
+static int _unit_file_enable(
```
i think the underscore prefix would be fine if we have functions which are wrapper by something else and that something else is really the only way to use the function. But that isn't the case here, it's used by multiple callers.

maybe rename this to unit_file_enable_now()? or so? or prefix with do_?

poettering commented on src/shared/install.c:
```diff
@@ -873,7 +890,7 @@ static int find_symlinks(
                 }
 
-                r = find_symlinks_in_directory(d, path, root_dir, i, match_name, ignore_same_name, config_path, same_name_link);
+                r = find_symlinks_in_directory(d, path, root_dir, i, true, match_name, ignore_same_name, config_path, same_name_link);
```
cryptic and long. maybe line beak? prefix the `true` with `/* ignore_destination= */`?

poettering commented on src/shared/install.c:
```diff
@@ -882,7 +899,7 @@ static int find_symlinks(
         /* We didn't find any suitable symlinks in .wants or .requires directories, let's look for linked unit files in this directory. */
         rewinddir(config_dir);
-        return find_symlinks_in_directory(config_dir, config_path, root_dir, i, match_name, ignore_same_name, config_path, same_name_link);
+        return find_symlinks_in_directory(config_dir, config_path, root_dir, i, false, match_name, ignore_same_name, config_path, same_name_link);
```
ditto

poettering commented on src/basic/unit-file.c:
```diff
@@ -361,7 +363,7 @@ int unit_file_resolve_symlink(
 
         *ret_destination = TAKE_PTR(dst);
-        return 0;
+        return tail ? false : true;  /* true if linked unit file */
```
!tail?

keszybz replied:
Unless I'm utterly confused, this is correct. If we have `tail`, then it's an alias, and this function is supposed to return false.

poettering replied:
yes i am just saying, why write such a complex ternary op, if `!tail` also works, and it is much shorter.

keszybz replied:
Oh, I see. I changed it like this now. I thought that the explicit form would be clearer, but guess not.

poettering commented:
what a PR. did a round of reviews. some points above. I think it would be good to split some parts out of this PR and merge independently.

I already cherry picked a bunch of commits into git main.

needs rebase hence.

keszybz commented:
Updated. I implemented the suggested changes, except as noted above. The only new stuff is commit "basic/stat-util: add null_or_empty_path_with_root()" near the beginning.

keszybz commented:
```
  ^-- SC2016 (info): Expressions don't expand in single quotes, use double quotes for that.
```
Fuck, that's why I added single quotes!

yuwata commented:
```
497/980 test-specifier                                                     FAIL             0.22s   killed by signal 6 SIGABRT
>>> SYSTEMD_LANGUAGE_FALLBACK_MAP=/root/src/src/locale/language-fallback-map MALLOC_PERTURB_=189 PATH=/root/build:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin /root/build/test-specifier
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
stderr:
/* test_specifier_escape */
/* test_specifier_escape_strv */
/* test_specifier_printf */
/* test_specifier_real_path */
Assertion 'r >= 0 || r == -EUNATCH' failed at src/test/test-specifier.c:102, function test_specifier_real_path(). Aborting.
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
```

keszybz commented:
ci/semaphoreci/pr: debian autopkgtest (lxc) failed with " unit-config          FAIL non-zero exit status 1 23:37 ". This is the same expected failure that was reported above.

bluca commented:
> Yeah, so this would need to be adjusted. But let's see if we actually want to do this before changing the tests there.
> 
> ```
> ======================================================================
> FAIL: test_unit_alias_enable (__main__.EnableTests)
> no sysv: enable unit with an alias
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/tmp/autopkgtest.OvTYXW/build.O10/systemd/debian/tests/unit-config", line 204, in test_unit_alias_enable
>     self.assertEqual(os.readlink(l),
> AssertionError: '../test_enable.service' != '/lib/systemd/system/test_enable.service'
> - ../test_enable.service
> + /lib/systemd/system/test_enable.service
> 
> 
> ======================================================================
> FAIL: test_unit_enable (__main__.EnableTests)
> no sysv: enable unit
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/tmp/autopkgtest.OvTYXW/build.O10/systemd/debian/tests/unit-config", line 105, in test_unit_enable
>     self.assertEqual(os.readlink(l),
> AssertionError: '../test_enable.service' != '/lib/systemd/system/test_enable.service'
> - ../test_enable.service
> + /lib/systemd/system/test_enable.service
> 
> 
> ======================================================================
> FAIL: test_unit_enable_full_path (__main__.EnableTests)
> systemctl enable a unit in a non-default path
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/tmp/autopkgtest.OvTYXW/build.O10/systemd/debian/tests/unit-config", line 356, in test_unit_enable_full_path
>     self.assertEqual(os.readlink(enable_l), f.name)
> AssertionError: '../tmpstnn9ebp.service' != '/tmp/tmpstnn9ebp.service'
> - ../tmpstnn9ebp.service
> ? ^^
> + /tmp/tmpstnn9ebp.service
> ? ^^^^
> 
> 
> ======================================================================
> FAIL: test_unit_sysv_alias_enable (__main__.EnableTests)
> with sysv: enable unit with an alias
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/tmp/autopkgtest.OvTYXW/build.O10/systemd/debian/tests/unit-config", line 249, in test_unit_sysv_alias_enable
>     self.assertEqual(os.readlink(l),
> AssertionError: '../test_enable.service' != '/lib/systemd/system/test_enable.service'
> - ../test_enable.service
> + /lib/systemd/system/test_enable.service
> 
> 
> ======================================================================
> FAIL: test_unit_sysv_enable (__main__.EnableTests)
> with sysv: enable unit
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/tmp/autopkgtest.OvTYXW/build.O10/systemd/debian/tests/unit-config", line 150, in test_unit_sysv_enable
>     self.assertEqual(os.readlink(l),
> AssertionError: '../test_enable.service' != '/lib/systemd/system/test_enable.service'
> - ../test_enable.service
> + /lib/systemd/system/test_enable.service
> ```

Updated:

https://salsa.debian.org/systemd-team/systemd/-/commit/e92d491d5af9304b93d0889bc2f3eb8604257c3f

keszybz commented:
centos ci (arch linux) – 20:44:09 TEST-46-HOMED

bluca commented on src/shared/tests.c:
```diff
@@ -132,6 +134,20 @@ int log_tests_skipped_errno(int r, const char *message) {
+        if (fd < 0)
+                return fd;
+
+        ssize_t l = strlen(contents);
```
strlen will bomb on null, so there should be at least an assert

bluca reviewed (changes requested):
> Since the user is allowed
> to move the unit file, or add a new unit file in a different location, and
> we don't actually follow the symlink, I think it's OK to create a dangling
> symlink. The prefix of "../" is useful to give a hint that the link points
> to files that are conceptually "one level up" in the directory hierarchy.

I think this is fine, but it needs to be documented clearly, as it will be quite unexpected to see dandling symlinks. We need to clarify in the manpage that this is an 'internal API' of sorts, and not supposed to point to the actual file. I don't think it's obvious for users.

keszybz commented:
> I think this is fine, but it needs to be documented clearly, as it will be quite unexpected to see dandling symlinks. We need to clarify in the manpage that this is an 'internal API' of sorts, and not supposed to point to the actual file. I don't think it's obvious for users.

I think this is known and expected, but I'll update the docs.

keszybz commented on test/test-systemctl-enable.sh:
```diff
@@ -0,0 +1,523 @@
+#!/usr/bin/env bash
+# SPDX-License-Identifier: LGPL-2.1-or-later
```
Yikes. I was calling the test by hand and forgot to hook it up.

keszybz commented:
Updated. I added the asserts, and hooked up the test, and added one more commit that tries to describe the alias/linking rules as they are.

keszybz commented:
> https://salsa.debian.org/systemd-team/systemd/-/commit/e92d491d5af9304b93d0889bc2f3eb8604257c3f

Thanks!

bluca commented on man/systemd.unit.xml:
```diff
@@ -175,11 +179,18 @@
+    source unit is a template, the target can also be a template, in which case the instance will be
+    "propagated" to the target unit to form a valid unit instance. The target of symlinks in
+    <filename>.wants/</filename> or <filename>.requires/</filename> must thus point to a valid unit file
+    location, i.e. the symlink target name must satisfy the described requirements, and the desination path
```
same typo

bluca commented:
Looks like the new test is failing here and there

keszybz commented:
This is new: CentOS CI (CentOS Stream 8):
```
12:08:27 [   60.674243] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
12:08:27 [   60.675765] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
12:08:27 [   60.676849] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
12:08:27 [   60.677478] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
12:08:27 [   60.677909] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
12:08:27 [   60.679033] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
12:08:27 [   60.680076] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
12:08:27 [   60.680146] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
12:08:27 [   60.779856] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
12:08:27 [   60.780586] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
12:08:27 [   60.781654] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
```
@yuwata, any idea?

keszybz commented:
I hope I fixed the tests. Test for %a is disabled for now.

yuwata commented:
> This is new: CentOS CI (CentOS Stream 8):
> ```
> 12:08:27 [   60.674243] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
> 12:08:27 [   60.675765] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
> 12:08:27 [   60.676849] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
> 12:08:27 [   60.677478] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
> 12:08:27 [   60.677909] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
> 12:08:27 [   60.679033] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
> 12:08:27 [   60.680076] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
> 12:08:27 [   60.680146] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
> 12:08:27 [   60.779856] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
> 12:08:27 [   60.780586] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
> 12:08:27 [   60.781654] systemd-udevd[634]: Assertion '!event_pid_changed(s->event)' failed at src/libsystemd/sd-event/sd-event.c:2585, function sd_event_source_set_enabled(). Ignoring.
> ```
> @yuwata, any idea?

Ouch. Will fix it tomorrow.

poettering commented on src/basic/unit-file.c:
```diff
@@ -395,77 +471,33 @@ int unit_file_build_name_map(
+                                if (!dst)
+                                        dst = strdup(filename);
+                                if (!dst)
+                                        return log_oom();
```
doesnt matter, but why not do:

```c
if (!dst) {
        dst = strdup(filename);
        if (!dst)
                return log_oom();
}
```

i think  that'd be a less suprising way to write this down, no?

poettering commented on src/basic/unit-file.c:
```diff
@@ -473,7 +505,7 @@ int unit_file_build_name_map(
                                 r = unit_file_resolve_symlink(lp->root_dir, lp->search_path,
                                                               *dir, dirfd(d), de->d_name,
-                                                              &dst);
+                                                              false, &dst);
```
doesn't matter, but maybe prefix "false" with a comment `/* resolve_destination_target= */` to make this easier to grok?

poettering commented on src/basic/os-util.c:
```diff
@@ -170,13 +170,19 @@ int open_extension_release(const char *root, const char *extension, char **ret_p
-                FOREACH_STRING(p, "/etc/os-release", "/usr/lib/os-release") {
-                        r = chase_symlinks(p, root, CHASE_PREFIX_ROOT,
+                const char *var = secure_getenv("SYSTEMD_OS_RELEASE");
+                if (var)
```
i wonder if it wouldn't be better to ignore the env var if root is set?

keszybz replied:
The opposite: I think it's particularly useful when operating with `--root`, because you might have os-release that is not set up yet, or is only partially filled in. It's also quite useful for testing.

poettering commented on src/test/test-specifier.c:
```diff
@@ -138,4 +138,18 @@ TEST(specifiers) {
+        assert_se(streq(resolved, "-----"));
+
+        assert_se(setenv("SYSTEMD_OS_RELEASE", "/nosuchfileordirectory", 1) == 0);
+        assert_se(specifier_printf("%A-%B-%M-%o-%w-%W", SIZE_MAX, specifier_table, NULL, NULL, &resolved) == -ENOENT);
```
maybe this behaviour isn't ideal. i.e. maybe we should treat a missing os-release file like an empty one in this context,

keszybz replied:
I thought about this, and that was my initial thought. But we generally expect os-release to be there. And if you're calling 'enable' on files with specifiers, I think it's more likely that something went wrong with the installation and os-release is missing (or in the wrong path or something) than that it's missing on purpose. We only throw the error if specifiers that refer to os-release are used. And it is OK to have an empty os-release file, that is tested in one of the tests. This also lightly pushes people towards installing os-release always, which I think is a reasonable thing to do.

poettering commented on src/shared/specifier.c:
```diff
@@ -270,37 +271,41 @@ int specifier_architecture(char specifier, const void *data, const char *root, c
+
+        /* Translate error for missing os-release file to EUNATCH. */
+        r = parse_os_release(root, id, ret);
+        return r == -ENOENT ? -EUNATCH : r;
```
so yeah, I have the suspicion we should just make this graceful, i.e. treat ENOENT like empty

poettering commented on src/shared/install.c:
```diff
@@ -865,7 +866,7 @@ static int find_symlinks(
 static int find_symlinks_in_scope(
                 UnitFileScope scope,
```
you can drop this param now? it's also available in info->scope?

poettering commented on src/shared/install.c:
```diff
@@ -1794,20 +1794,22 @@ static int install_info_symlink_alias(
                 if (q < 0) {
                         unit_file_changes_add(changes, n_changes, q, *s, NULL);
-                        return q;
+                        r = r < 0 ? r : q;
```
one day we should add a macro for this, i.e. `r = ERRNO_COLLECT(r, q)` or so

poettering commented on src/basic/unit-file.c:
```diff
@@ -111,6 +112,11 @@ int unit_validate_alias_symlink_or_warn(int log_level, const char *filename, con
         /* dst checks */
 
+        if (streq(src, dst))
+                return log_debug_errno(SYNTHETIC_ERRNO(ELOOP),
```
shouldn't we be careful with ELOOP? might also happen for other reasons, no?

poettering commented on src/shared/install.c:
```diff
@@ -2702,17 +2702,66 @@ int unit_file_disable(
+
+                r = path_extract_filename(*a, &n);
+                if (r < 0)
+                        return r;
```
btw, it think we should generate an error if path_extrct_filename() returns O_DIRECTORY. (it does that in case the path is suffixed by `/`). It returns 0 if the path is not suffixed like that. And here we only care about files, hence we should refuse such paths to dirs.

poettering commented on src/shared/install.c:
```diff
@@ -2702,17 +2702,66 @@ int unit_file_disable(
+                if (i && i->path)
+                        /* Use startswith here, because we know that paths are normalized, and
+                         * path_startswith() would give us a relative path, but we need an absolute path
+                         * relative to i->root. */
```
an absolute paths that is relative? waahhhh? not following?

this funciton is supposed to normalize, but the paths already *are* normalized?

not following here...

keszybz replied:
I extended the comment.

poettering commented on src/shared/install.c:
```diff
@@ -1606,7 +1606,8 @@ static int install_info_traverse(
 
-                r = install_info_follow(ctx, i, lp, flags, false);
+                r = install_info_follow(ctx, i, lp, flags,
+                                        i->type == UNIT_FILE_TYPE_LINKED);  /* If linked, don't look at the target name */
```
as usual, I'd prefix this with the parameter name in a comment

poettering commented on src/shared/install.c:
```diff
@@ -423,21 +423,54 @@ void unit_file_dump_changes(int r, const char *verb, const UnitFileChange *chang
+        int r;
+
+        if (!path_is_absolute(target_a) || !path_is_absolute(target_b)) {
+                dirname = dirname_malloc(src);
```
path_extract_directory()

poettering commented on src/shared/install.c:
```diff
@@ -423,21 +423,54 @@ void unit_file_dump_changes(int r, const char *verb, const UnitFileChange *chang
+        }
+
+        _cleanup_free_ char *a = strjoin(path_is_absolute(target_a) ? root : dirname, "/", target_a);
+        _cleanup_free_ char *b = strjoin(path_is_absolute(target_b) ? root : dirname, "/", target_b);
```
path_join()?

(yes, the paths might be normalized, but so what, this is not performance critical. And it is shorter, and easier to read)

yuwata commented on src/basic/unit-file.c:
```diff
@@ -259,27 +259,45 @@ static int directory_name_is_valid(const char *name) {
                                          dir, dir ? "/" : "", filename);
 
+        if (!dir) {
+                assert(is_path(filename));
```
`path_is_absolute()`?? Moreover, this is not necessary, as it is checked at the beginning of this function.

yuwata commented on src/basic/unit-file.c:
```diff
@@ -259,27 +259,45 @@ static int directory_name_is_valid(const char *name) {
+                        return r;
+
+                dir = _dir;
+                filename = basename(filename);
```
`path_extract_filename()`

keszybz commented:
Updated. I think I addressed all the comments. I few things that were questions/suggestions are not done, but I think it's OK to leave them for later. I replied on the comments. Taking liberty to restore the green label.

keszybz commented:
It's been running for 5+ h. I'll merge this without waiting for focal-*.

buczek commented:
~~Did this change the semantic of symlink chains like `/etc/systemd/system/default.target` -> `/etc/local/systemd_system_default.target` -> ` /lib/systemd/system/graphical.target` ? It looks like in v242 such a chain would be interpreted as an alias and in v260.2 it would be interpreted as a "linked unit file" with the result that dependencies of graphical.target are ignored. I didn't analyze the code in this PR, but I guess, this change is in here. Is that correct?~~

~~One could argue that the new behavior matches systemd.unit(5), but on the other hand, the man page itself was only updated regarding this in this PR (ecd6c000d3a2e7). It is a surprise and broke our systems.~~

Edit: Sorry, I guess the relevant change was in e8630e6952 ("pid1: use a cache for all unit aliases") and not in this PR.
